docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,62 @@
+The purpose of this file is to describe the functional and key architecture of
+this project. It is spiced with some rational throughout.
+
+# Slots & Tokens
+
+This library shall provide multiple tokens for the user(s). Any given slot can
+only contain one token at any given time. In case of smartcards the slot is a
+reader whilst the token is the actual smartcard. Since this project won't
+implement a separate tool for "virtual swapping" of tokens, each token is
+associated with its own slot. Thus, the library offers a set of slots with a
+token inside each of these.
+
+In order to allow the creation of new tokens using the pkcs11 means, this
+library presents one more slot/token pair than tokens currently exist. The
+additional token is a blank token that can be initialized using the C_InitToken
+function.
+(Note: This is not implemented yet)
```

Comment:
No that's how SoftHSM works. You can use the ptool command "addtoken" to add an uninitialized token if you wanted too.

Reply:
That was my thinking on how to do things after "rearranging" some of the code.
Nice this would be that we then don't need any external tools anymore.
Or would you say that such an empty token would be too confusing and we should only allow adding of tokens via an extra command ?

Reply:
I don't think it's super confusing, as SoftHSM does it. But I think it's odd. As you create tokens a new one keeps existing.

I don't think you ever get away from external commands. The PKCS#11 spec isn't 100% functional for card initialization. Offhand I forget where it gets weird, but everything always ships a provisioning tool, even SoftHSM

Reply:
I guess creating a token with an SO-PIN only can be done with a single future-api call. Adding the user-PIN via the SO-PIN and then creating the key object would then be, where PKCS11 functions take over ?

Reply:
@AndreasFuchsSIT I think they may actually just be limitations on the actual card side